### PR TITLE
fix issue 16098 - Occasional runtime segfault of programs with SIMD 32 bytes

### PR DESCRIPTION
**Pull request by SSoulaimane**
The cause of these occasional failures is misalignment.

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]
I can hear water now saying "those `bool`s don't change, make 'em const!"

Otherwise LGTM.

**Comment by SSoulaimane** [comment]
> I can hear water now saying "those bools don't change, make 'em const!"

That's true. I've been taken by the sloppiness and the C-ish nature of the backend.

**Comment by thewilsonator** [comment]
Yep, it used to be C.

**Comment by SSoulaimane** [comment]
As a comment. As an alternative to this general pessimistic approach We can at least for stack variables force the stack to be aligned to a 32 bytes boundary. LDC generates an instruction in the prolog of any function which has SIMD 32 bytes local variables.
The instruction looks like the following:
```asm
and    rsp,0xffffffffffffffe0     # 0xffffffffffffffe0 == ~(32 - 1)
```
This strategy may bring performance improvements.

**Comment by thewilsonator** [comment]
Nice idea, although DMD should probably focus on being correct more than it should being fast. If its not to hard to do it, by all means go ahead. Also if its possible to do a test this that would also be great.

**Comment by WalterBright** [comment]
I don't really understand how the stack is getting misaligned. `floatreg` is supposed to be always aligned. Assuming the stack is not aligned means performance will be degraded.

**Comment by SSoulaimane** [comment]
> I don't really understand how the stack is getting misaligned. `floatreg` is supposed to be always aligned. Assuming the stack is not aligned means performance will be degraded.

This is somewhat related to issue [16098](https://issues.dlang.org/show_bug.cgi?id=16098). The problem is that the stack alignment is < 32, and if you assume that it will be  somehow aligned to 32 bytes and use an aligned instruction (for example `vmovapd`) instead of an unaligned one (`vmovupd`) then the CPU may crash your program.

**Comment by SSoulaimane** [comment]
> floatreg is supposed to be always aligned.

Yes that's true. I fixed that.

**Comment by WalterBright** [comment]
> The problem is that the stack alignment is < 32

That should never happen with code generation that needs 32 bit alignment, and the compiler should always be emitting code that keeps the stack so aligned. (The code gen may skip such alignment if the function 1. does not need the alignment and 2. transitively does not call functions that need the alignment.)

On a pragmatic note, I wonder how the compiler could have persisted for a decade generating a misaligned stack without a blizzard of bug reports :-) I know there were some in earlier days, but they were all quickly corrected.

**Comment by ZombineDev** [comment]
@WalterBright I suppose the issue is with 32 **byte** (256-bit) alignment, not "32-bit".

**Comment by WalterBright** [comment]
@ZombineDev You're right.

@SSoulaimane how does LDC deal with pointer indirections? Always assume misaligned (as this PR does)?

**Comment by thewilsonator** [comment]
> how does LDC deal with pointer indirections?

Defaults to the alignment of the type, you can use an unaligned with https://github.com/ldc-developers/druntime/blob/ldc/src/ldc/simd.di#L196

**Comment by WalterBright** [comment]
Makes sense. Looks like we have little choice but to get the stack aligned for 32 byte values.

**Comment by SSoulaimane** [comment]
> Makes sense. Looks like we have little choice but to get the stack aligned for 32 byte values.

I'm working on it right now. It's a bit more work than just enforcing the new alignment because now we can no longer use `rbp` instead we need to use `rsp`. It will need a review once It's done.

**Comment by SSoulaimane** [comment]
> > how does LDC deal with pointer indirections?
> 
> Defaults to the alignment of the type, you can use an unaligned with https://github.com/ldc-developers/druntime/blob/ldc/src/ldc/simd.di#L196

Apparently so does C++. Using malloc for SIMD is unsupported. So we can get of that pessimism.

**Comment by SSoulaimane** [comment]
> Doesn't this mean that two frame pointers are required? I suppose one could use EBP to access the non-aligned data, and ESP for the aligned data.

I don't know how that can work.

**Comment by WalterBright** [comment]
> I don't know how that can work.

I don't know how it can work otherwise. The function prolog starts with parameters and the return address pushed on the stack. RSP is pointing to it. Now the RSP gets aligned by adding 0..15 bytes to it. How does one then access the parameters and return value, if not the original RSP value being saved (presumably in RBP)? But that RBP cannot be used to access the aligned data, because it will be off by a 0..15 runtime, not compile time, value. But RSP can be used to access that aligned data.

Hence both RBP and RSP being used as frame pointers. RBP for the parameters and return value, RSP for the locals.

**Comment by SSoulaimane** [comment]
> ... RBP for the parameters and return value, RSP for the locals.

That's what I ended up doing.

> BTW, in order to make PRs more comprehensible, changing the stack alignment scheme should be done as a separate PR.

What do you mean?

**Comment by WalterBright** [comment]
@klickverbot thanks for the info confirming my supposition.

**Comment by WalterBright** [comment]
> That's what I ended up doing.

So I see.

> What do you mean?

Sorry, I thought this PR included code about loading vectors. I was mistaken.

File: src/dmd/backend/cgcod.d [hunk 1]
```diff
@@ -222,6 +223,7 @@ tryagain:
 
     // if no parameters, assume we don't need a stack frame
     needframe = 0;
+    enforcealign = 0;
```
Comment (WalterBright) [inline src/dmd/backend/cgcod.d]: being a bool, this should be false, not 0

File: src/dmd/backend/cgcod.d [hunk 2]
```diff
@@ -305,6 +308,37 @@ tryagain:
         regcon.params &= ~noparams;
     }
 
+    // See if we need to enforce a particular stack alignment
+    for (int i = 0; i < globsym.top; i++)
+    {
+        Symbol *s = globsym.tab[i];
```
Comment (WalterBright) [inline src/dmd/backend/cgcod.d]:
```
foreach (i; 0 .. globsym.top)
```

File: src/dmd/backend/cgcod.d [hunk 3]
```diff
@@ -305,6 +308,37 @@ tryagain:
...
+        const sz = type_alignsize(s.Stype);
+
+        switch (s.Sclass)
+        {
+            case SCregister:
+            case SCauto:
+            case SCfastpar:
+                if (s.Sfl == FLreg)
+                    break;
+
```
Comment (WalterBright) [inline src/dmd/backend/cgcod.d]: move line 320 here

File: src/dmd/backend/cgcod.d [hunk 4]
```diff
@@ -865,6 +899,9 @@ Lagain:
     */
 
     int bias = cast(int)(Para.size + (needframe ? 0 : REGSIZE));
+    if (enforcealign)
+        bias = 0; // bias is unknown
```
Comment (WalterBright) [inline src/dmd/backend/cgcod.d]: I'd rather merge 901..903 into a ?: statement and then make `bias` a `const`.

File: src/dmd/backend/cgcod.d [hunk 5]
```diff
@@ -937,14 +974,11 @@ Lagain:
...
-        if (STACKALIGN == 16)
-        {
-            if (sz & (8|4))
-                localsize += STACKALIGN - (sz & (8|4));
-        }
-        else if (sz & 4)
-            localsize += 4;
+        int sz = cast(int)(localsize + npush * REGSIZE);
+        if (!enforcealign)
+            sz += Para.size + (needframe ? 0 : -REGSIZE);
```
Comment (WalterBright) [inline src/dmd/backend/cgcod.d]: Same ?: and const as above.

File: src/dmd/backend/cod1.d [hunk 6]
```diff
@@ -3078,6 +3078,11 @@ void cdfunc(ref CodeBuilder cdb, elem* e, regm_t* pretregs)
         if (alignsize > stackalign &&
             (I64 || (alignsize == 16 && tyvector(ep.Ety))))
         {
+            if (tyvector(ep.Ety) && alignsize > STACKALIGN)
```
Comment (WalterBright) [inline src/dmd/backend/cod1.d]: I don't think the tyvector() clause is necessary.

File: src/dmd/backend/cod1.d [hunk 7]
```diff
@@ -3078,6 +3078,11 @@ void cdfunc(ref CodeBuilder cdb, elem* e, regm_t* pretregs)
...
+            {
+                STACKALIGN = alignsize;
+                enforcealign = 1;
```
Comment (WalterBright) [inline src/dmd/backend/cod1.d]: `= true;`

File: src/dmd/backend/cod3.d [hunk 8]
```diff
@@ -4789,20 +4820,34 @@ void assignaddrc(code *c)
...
                     if (c.Irm & 8)
                         c.Irex |= REX_R;
                     c.Irm = modregrm(2,SP,BP);
                     c.Iflags = CFoff;
                     c.IFL1 = FLconst;
-                    c.IEV1.Vuns = -EBPtoESP - REGSIZE;
+                    c.IEV1.Vuns = -EBPtoESP;
+                    if (enforcealign)
+                    {
+                        // AND ESP, -STACKALIGN
```
Comment (WalterBright) [inline src/dmd/backend/cod3.d]: This seems redundant with line 531?

**Comment by WalterBright** [comment]
@SSoulaimane this is pretty nice work! Nice to see someone figure out how the back end works!

**Comment by SSoulaimane** [comment]
> Nice to see someone figure out how the back end works!

Enjoyed every bit of it.

File: src/dmd/backend/cod1.d (hunk 6 above)
Comment (SSoulaimane) [reply src/dmd/backend/cod1.d]: I64 includes user defined types.

File: src/dmd/backend/cod3.d (hunk 8 above)
Comment (WalterBright) [reply src/dmd/backend/cod3.d]: Yes, I just hate to see the duplication.

File: src/dmd/backend/cod1.d (hunk 6 above)
Comment (WalterBright) [reply src/dmd/backend/cod1.d]: Yes, but if their alignsize is larger than STACKALIGN, shouldn't they get so aligned, too? After all, what if the vector type is a field in a struct?

Comment (WalterBright) [reply src/dmd/backend/cod1.d]: or an array of vectors?

Comment (SSoulaimane) [reply src/dmd/backend/cod1.d]: True.

**Comment by thewilsonator** [comment]
Let me know when this is good to go.

**Comment by WalterBright** [comment]
It's failing consistently on OSX. Otherwise, looks good.

**Comment by thewilsonator** [comment]
Correction, it fails on OSX32, can we just rip support for it already?

**Comment by WalterBright** [comment]
I kinda hate to do that without knowing why it is failing. It may be a latent bug on the other platforms.

**Comment by thewilsonator** [comment]
>I kinda hate to do that without knowing why it is failing. It may be a latent bug on the other platforms.

I know how you feel, OTOH, this is a rather serious bug and more importantly is required for the rest of the great work @SSoulaimane has been doing, which is crucial for the support and interaction of C/C++.

**Comment by thewilsonator** [comment]
```
lldb /Users/nicholaswilson/d/phobos/generated/osx/release/32/tests_extractor
(lldb) target create "/Users/nicholaswilson/d/phobos/generated/osx/release/32/tests_extractor"
Current executable set to '/Users/nicholaswilson/d/phobos/generated/osx/release/32/tests_extractor' (i386).
(lldb) r
Process 84224 launched: '/Users/nicholaswilson/d/phobos/generated/osx/release/32/tests_extractor' (i386)
Process 84224 stopped
* thread #1: tid = 0x35954b, 0x00011594 tests_extractor`_D3std6getopt__TQkTEQsQq6configTAyaTQeTPQiTQlTQoTQkTQuTQxTQtTQBdTQBhTQBeTQBpTQBtTPbZQDaFNfKAQCiQCyQCoQCrQCnQCxQDaQCwQDgQDjQDfQDpQDsQDoQDyQEbQChZSQFnQFm12GetoptResult + 32, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0xa)
    frame #0: 0x00011594 tests_extractor`_D3std6getopt__TQkTEQsQq6configTAyaTQeTPQiTQlTQoTQkTQuTQxTQtTQBdTQBhTQBeTQBpTQBtTPbZQDaFNfKAQCiQCyQCoQCrQCnQCxQDaQCwQDgQDjQDfQDpQDsQDoQDyQEbQChZSQFnQFm12GetoptResult + 32
tests_extractor`_D3std6getopt__TQkTEQsQq6configTAyaTQeTPQiTQlTQoTQkTQuTQxTQtTQBdTQBhTQBeTQBpTQBtTPbZQDaFNfKAQCiQCyQCoQCrQCnQCxQDaQCwQDgQDjQDfQDpQDsQDoQDyQEbQChZSQFnQFm12GetoptResult:
->  0x11594 <+32>: pushl  (%edx)
    0x11596 <+34>: movl   -0xc(%ebp), %ebx
    0x11599 <+37>: movl   0x155aa6(%ebx), %ecx
    0x1159f <+43>: pushl  %ecx
(lldb) bt
* thread #1: tid = 0x35954b, 0x00011594 tests_extractor`_D3std6getopt__TQkTEQsQq6configTAyaTQeTPQiTQlTQoTQkTQuTQxTQtTQBdTQBhTQBeTQBpTQBtTPbZQDaFNfKAQCiQCyQCoQCrQCnQCxQDaQCwQDgQDjQDfQDpQDsQDoQDyQEbQChZSQFnQFm12GetoptResult + 32, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0xa)
  * frame #0: 0x00011594 tests_extractor`_D3std6getopt__TQkTEQsQq6configTAyaTQeTPQiTQlTQoTQkTQuTQxTQtTQBdTQBhTQBeTQBpTQBtTPbZQDaFNfKAQCiQCyQCoQCrQCnQCxQDaQCwQDgQDjQDfQDpQDsQDoQDyQEbQChZSQFnQFm12GetoptResult + 32
    frame #1: 0x00002890 tests_extractor`_Dmain + 416
    frame #2: 0x00042ff3 tests_extractor`_D2rt6dmain211_d_run_mainUiPPaPUAAaZiZ6runAllMFZ9__lambda1MFZv + 27
    frame #3: 0x00042e65 tests_extractor`_D2rt6dmain211_d_run_mainUiPPaPUAAaZiZ7tryExecMFMDFZvZv + 21
    frame #4: 0x00042f6d tests_extractor`_D2rt6dmain211_d_run_mainUiPPaPUAAaZiZ6runAllMFZv + 173
    frame #5: 0x00042e65 tests_extractor`_D2rt6dmain211_d_run_mainUiPPaPUAAaZiZ7tryExecMFMDFZvZv + 21
    frame #6: 0x00042ddc tests_extractor`_d_run_main + 516
    frame #7: 0x00002e3a tests_extractor`main + 50
    frame #8: 0x000026e5 tests_extractor`start + 53
(lldb) dis
tests_extractor`_D3std6getopt__TQkTEQsQq6configTAyaTQeTPQiTQlTQoTQkTQuTQxTQtTQBdTQBhTQBeTQBpTQBtTPbZQDaFNfKAQCiQCyQCoQCrQCnQCxQDaQCwQDgQDjQDfQDpQDsQDoQDyQEbQChZSQFnQFm12GetoptResult:
    0x11574 <+0>:   pushl  %ebp
    0x11575 <+1>:   movl   %esp, %ebp
    0x11577 <+3>:   subl   $0x28, %esp
    0x1157a <+6>:   movl   %esi, -0x14(%ebp)
    0x1157d <+9>:   movl   %eax, %esi
    0x1157f <+11>:  movl   %ebx, -0x18(%ebp)
    0x11582 <+14>:  movl   %edi, -0x10(%ebp)
    0x11585 <+17>:  calll  0x1158a                   ; <+22>
    0x1158a <+22>:  popl   %ecx
    0x1158b <+23>:  movl   %ecx, -0xc(%ebp)
    0x1158e <+26>:  subl   $0xc, %esp
    0x11591 <+29>:  movl   0x70(%ebp), %edx
->  0x11594 <+32>:  pushl  (%edx)
    0x11596 <+34>:  movl   -0xc(%ebp), %ebx
    0x11599 <+37>:  movl   0x155aa6(%ebx), %ecx
    0x1159f <+43>:  pushl  %ecx
    0x115a0 <+44>:  xorl   %edi, %edi
    0x115a2 <+46>:  pushl  %edi
    0x115a3 <+47>:  movl   -0xc(%ebp), %ebx
    0x115a6 <+50>:  leal   0xf3873(%ebx), %ecx
    0x115ac <+56>:  pushl  %ecx
    0x115ad <+57>:  movl   $0x1ac, %eax              ; imm = 0x1AC
    0x115b2 <+62>:  pushl  $0xc
    0x115b4 <+64>:  calll  0xfb59a                   ; symbol stub for: _D3std9exception__T7enforceZ__TQmTkZQrFNaNfkLAxaAyakZk
    0x115b9 <+69>:  addl   $0xc, %esp
    0x115bc <+72>:  movl   %esi, %edx
    0x115be <+74>:  movb   $0x0, -0x4(%ebp)
    0x115c2 <+78>:  xorl   %ebx, %ebx
    0x115c4 <+80>:  leal   -0x4(%ebp), %edi
    0x115c7 <+83>:  movl   %ebx, (%edx)
    0x115c9 <+85>:  leal   -0x8(%ebp), %ecx
    0x115cc <+88>:  movl   %ebx, 0x4(%edx)
    0x115cf <+91>:  movl   %ebx, 0x8(%edx)
    0x115d2 <+94>:  movl   %ebx, -0x8(%ebp)
    0x115d5 <+97>:  subl   $0x4, %esp
    0x115d8 <+100>: pushl  0x70(%ebp)
    0x115db <+103>: pushl  %edi
    0x115dc <+104>: pushl  %edx
    0x115dd <+105>: pushl  %ecx
    0x115de <+106>: pushl  %ebx
    0x115df <+107>: pushl  %ebx
    0x115e0 <+108>: pushl  0x6c(%ebp)
    0x115e3 <+111>: pushl  0x68(%ebp)
    0x115e6 <+114>: pushl  0x64(%ebp)
    0x115e9 <+117>: pushl  0x60(%ebp)
    0x115ec <+120>: pushl  0x5c(%ebp)
    0x115ef <+123>: subl   $0x4, %esp
    0x115f2 <+126>: pushl  0x58(%ebp)
    0x115f5 <+129>: pushl  0x54(%ebp)
    0x115f8 <+132>: pushl  0x50(%ebp)
    0x115fb <+135>: pushl  0x4c(%ebp)
    0x115fe <+138>: pushl  0x48(%ebp)
    0x11601 <+141>: subl   $0x4, %esp
    0x11604 <+144>: pushl  0x44(%ebp)
    0x11607 <+147>: pushl  0x40(%ebp)
    0x1160a <+150>: pushl  0x3c(%ebp)
    0x1160d <+153>: pushl  0x38(%ebp)
    0x11610 <+156>: pushl  0x34(%ebp)
    0x11613 <+159>: subl   $0x4, %esp
    0x11616 <+162>: pushl  0x30(%ebp)
    0x11619 <+165>: pushl  0x2c(%ebp)
    0x1161c <+168>: pushl  0x28(%ebp)
    0x1161f <+171>: pushl  0x24(%ebp)
    0x11622 <+174>: pushl  0x20(%ebp)
    0x11625 <+177>: subl   $0x4, %esp
    0x11628 <+180>: movl   0x8(%ebp), %eax
    0x1162b <+183>: pushl  0x1c(%ebp)
    0x1162e <+186>: pushl  0x18(%ebp)
    0x11631 <+189>: pushl  0x14(%ebp)
    0x11634 <+192>: pushl  0x10(%ebp)
    0x11637 <+195>: pushl  0xc(%ebp)
    0x1163a <+198>: calll  0xfb1d4                   ; symbol stub for: _D3std6getopt__T10getoptImplTEQBcQBb6configTAyaTQeTPQiTQlTQoTQkTQuTQxTQtTQBdTQBhTQBeTQBpTQBtTPbZQDcFNfKAQCiKSQEdQEc13configurationKSQFaQEz12GetoptResultKCQFwQFv15GetOptExceptionHQFeAvQgQGaQFoQFrQFnQFxQGaQFwQGgQGjQGfQGpQGsQGoQGyQHbQFhZv
    0x1163f <+203>: addl   $0x4, %esp
    0x11642 <+206>: movsbl (%esi), %edx
    0x11645 <+209>: xorb   $0x1, %dl
    0x11648 <+212>: je     0x11669                   ; <+245>
    0x1164a <+214>: cmpl   %ebx, -0x8(%ebp)
    0x1164d <+217>: je     0x11669                   ; <+245>
    0x1164f <+219>: movl   -0x8(%ebp), %ebx
    0x11652 <+222>: movl   %ebx, -0x28(%ebp)
    0x11655 <+225>: calll  0xfc20c                   ; symbol stub for: _d_throwdwarf
    0x1165a <+230>: movl   -0x18(%ebp), %ebx
    0x1165d <+233>: movl   -0x14(%ebp), %esi
    0x11660 <+236>: movl   -0x10(%ebp), %edi
    0x11663 <+239>: movl   %ebp, %esp
    0x11665 <+241>: popl   %ebp
    0x11666 <+242>: retl   $0x6c
    0x11669 <+245>: movl   %esi, %eax
    0x1166b <+247>: movl   -0x18(%ebp), %ebx
    0x1166e <+250>: movl   -0x14(%ebp), %esi
    0x11671 <+253>: movl   -0x10(%ebp), %edi
    0x11674 <+256>: movl   %ebp, %esp
    0x11676 <+258>: popl   %ebp
    0x11677 <+259>: retl   $0x6c
    0x1167a <+262>: nop
(lldb) register read
General Purpose Registers:
       eax = 0xbffff660
       ebx = 0x01805001
       ecx = 0x0001158a  tests_extractor`_D3std6getopt__TQkTEQsQq6configTAyaTQeTPQiTQlTQoTQkTQuTQxTQtTQBdTQBhTQBeTQBpTQBtTPbZQDaFNfKAQCiQCyQCoQCrQCnQCxQDaQCwQDgQDjQDfQDpQDsQDoQDyQEbQChZSQFnQFm12GetoptResult + 22
       edx = 0x0000000a
       edi = 0x00002710  tests_extractor`_Dmain + 32
       esi = 0xbffff660
       ebp = 0xbffff528
       esp = 0xbffff4f4
        ss = 0x00000023
    eflags = 0x00210292
       eip = 0x00011594  tests_extractor`_D3std6getopt__TQkTEQsQq6configTAyaTQeTPQiTQlTQoTQkTQuTQxTQtTQBdTQBhTQBeTQBpTQBtTPbZQDaFNfKAQCiQCyQCoQCrQCnQCxQDaQCwQDgQDjQDfQDpQDsQDoQDyQEbQChZSQFnQFm12GetoptResult + 32
        cs = 0x0000001b
        ds = 0x00000023
        es = 0x00000023
        fs = 0x00000000
        gs = 0x0000000f
```

**Comment by SSoulaimane** [comment]
> (above is an lldb session with said code)

Thank you. That was perfect.

**Comment by SSoulaimane** [comment]
We're good. I think.

**Comment by jacob-carlborg** [comment]
So the fix for macOS 32bit was https://github.com/dlang/dmd/pull/9143/commits/5bf961203a137c3f40be8e679aed30a0b6c22f21 ?

**Comment by SSoulaimane** [comment]
> So the fix for macOS 32bit was [5bf9612](https://github.com/dlang/dmd/commit/5bf961203a137c3f40be8e679aed30a0b6c22f21) ?

Yes.

**Pull request merged**